### PR TITLE
Fix Windows unit test failing because of DataStore threading issue

### PR DESCRIPTION
**Pull request by SimonMarquis**
This PR fixes issues on Windows environments, where DataStore threading [is known to be buggy](https://issuetracker.google.com/issues/203087070).
A fix has been awaited for months now, but still remains...
Replacing the default DataStore implementation with an in-memory fake avoids having to deal with file synchronization that lead to the issue.

Here is an example of such errors that appears on Windows env:
```
> Task :core:datastore:testDemoDebugUnitTest

com.google.samples.apps.nowinandroid.core.datastore.NiaPreferencesDataSourceTest > userShouldHideOnboarding_unfollowsLastTopic_shouldHideOnboardingIsFalse FAILED
    java.io.IOException: Unable to rename C:\Users\RUNNER~1\AppData\Local\Temp\junit11942129073694190937\user_preferences_test.pb.tmp.This likely means that there are multiple instances of DataStore for this file. Ensure that you are only creating a single instance of datastore for this file.
```

It fixes both of these issues that can be closed:

- https://github.com/android/nowinandroid/issues/98
- https://github.com/android/nowinandroid/issues/1242

Here is a before/after (red/green) on a Windows env in GHA:
- https://github.com/SimonMarquis/nowinandroid/pull/21/commits

### CODE REVIEW

**Comment by SimonMarquis** [comment]
@dturner I don't know exactly how many Windows users there are in the wild building this project but I'm pretty sure it's not zero (I'm one of them). It'd great to fix unit tests for them :)

**Comment by lihenggui** [comment]
> @dturner I don't know exactly how many Windows users there are in the wild building this project but I'm pretty sure it's not zero (I'm one of them). It'd great to fix unit tests for them :)

It is definitely not zero. I am one of the developers using Windows to build this project. : )

File: core/datastore/src/test/kotlin/com/google/samples/apps/nowinandroid/core/datastore/NiaPreferencesDataSourceTest.kt
Comment (Jaehwa-Noh) [inline core/datastore/src/test/kotlin/com/google/samples/apps/nowinandroid/core/datastore/NiaPreferencesDataSourceTest.kt:1]: I concerned that this doesn't test real PreferencesDataStore, Is this tests valuable?

**Review by Jaehwa-Noh — COMMENTED** [review]
image

Congratulation! Works on Window!

File: core/datastore/src/test/kotlin/com/google/samples/apps/nowinandroid/core/datastore/NiaPreferencesDataSourceTest.kt
Comment (SimonMarquis) [reply core/datastore/src/test/kotlin/com/google/samples/apps/nowinandroid/core/datastore/NiaPreferencesDataSourceTest.kt:1]: which test exactly?

Comment (Jaehwa-Noh) [reply core/datastore/src/test/kotlin/com/google/samples/apps/nowinandroid/core/datastore/NiaPreferencesDataSourceTest.kt:1]: All tests in this file.

Comment (SimonMarquis) [reply core/datastore/src/test/kotlin/com/google/samples/apps/nowinandroid/core/datastore/NiaPreferencesDataSourceTest.kt:1]: This class is testing `NiaPreferencesDataSource`, not the `DataStore<UserPreferences>` implementation.

**Comment by SimonMarquis** [comment]
@dturner PTAL 🙏

**Review by dturner — APPROVED** [review]

**Comment by SimonMarquis** [comment]
I'll fix the badging issue in another PR.

**Review by dturner — APPROVED** [review]
Thanks for doing this. In memory datastore should also result in faster tests 👍

**Pull request merged**